MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -1786,28 +1896,88 @@ GetMemoryMapWithPopulatedAccessAttributes (
     DumpImageRecords (&mImagePropertiesPrivate.ImageRecordList);
     );
 
-  // Split PE code/data if firmware volume image protection is active
-  if (gDxeMps.ImageProtectionPolicy.Fields.ProtectImageFromFv) {
-    SeparateImagesInMemoryMap (MemoryMapSize, *MemoryMap, *DescriptorSize, &mImagePropertiesPrivate.ImageRecordList, AdditionalRecordCount);
+  // Set MergedImageList to the list of IMAGE_PROPERTIES_RECORD entries which will
+  // be used to breakup the memory map
+  if (mImagePropertiesPrivate.ImageRecordCount == 0) {
+    MergedImageList = &mNonProtectedImageRangesPrivate.NonProtectedImageList;
+  } else if (mNonProtectedImageRangesPrivate.NonProtectedImageCount == 0) {
+    MergedImageList = &mImagePropertiesPrivate.ImageRecordList;
+  } else {
+    MergedImageList          = &mImagePropertiesPrivate.ImageRecordList;
+    ArrayOfListEntryPointers = AllocateZeroPool (mNonProtectedImageRangesPrivate.NonProtectedImageCount * sizeof (LIST_ENTRY *));
+
+    if (ArrayOfListEntryPointers == NULL) {
+      FreePool (*MemoryMap);
+      *MemoryMapSize  = 0;
+      *DescriptorSize = 0;
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    Status = MergeImagePropertiesRecordLists (
+               MergedImageList,
+               &mNonProtectedImageRangesPrivate.NonProtectedImageList,
+               mNonProtectedImageRangesPrivate.NonProtectedImageCount,
+               ArrayOfListEntryPointers
+               );
+
+    ASSERT_EFI_ERROR (Status);
   }
 
-  DEBUG_CODE (
-    DEBUG ((DEBUG_INFO, "---Memory Map With Separated Image Descriptors---\n"));
-    DumpMemoryMap (MemoryMapSize, *MemoryMap, DescriptorSize);
-    DEBUG ((DEBUG_INFO, "---------------------------------------\n"));
-    );
+  SeparateImagesInMemoryMap (MemoryMapSize, *MemoryMap, *DescriptorSize, MergedImageList, AdditionalRecordCount);
 
   NumberOfDescriptors   = *MemoryMapSize / *DescriptorSize;
   NumberOfBitmapEntries = (NumberOfDescriptors % 8) == 0 ? NumberOfDescriptors : (((NumberOfDescriptors / 8) * 8) + 8);
 
   Bitmap = AllocateZeroPool (NumberOfBitmapEntries / 8);
 
+  if (Bitmap == NULL) {
+    FreePool (*MemoryMap);
+    *MemoryMapSize  = 0;
+    *DescriptorSize = 0;

```

Comment:
I think there are some memory leaks in the error conditions.

ArrayOfListEntryPointers

Reply:
Added an OutOfResourcesCleanup section with gotos whenever an allocation fails and also free ArrayOfListEntryPointers in the non-error case.